tensorflow/c/experimental/filesystem/plugins/cache/expiring_lru_cache.h

```diff
@@ -0,0 +1,201 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_C_EXPERIMENTAL_FILESYSTEM_PLUGINS_CACHE_EXPIRING_LRU_CACHE_H_
+#define TENSORFLOW_C_EXPERIMENTAL_FILESYSTEM_PLUGINS_CACHE_EXPIRING_LRU_CACHE_H_
+
+#include <chrono>
+#include <list>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+
+#include "tensorflow/c/experimental/filesystem/plugins/cache/thread_annotations.h"
+#include "tensorflow/c/tf_status.h"
+
+namespace tf_cache {
+
+namespace {
+typedef std::function<uint64_t()> TimerSeconds;
+uint64_t DefaultTimer() {
+  return std::chrono::duration_cast<std::chrono::seconds>(
+             std::chrono::high_resolution_clock::now().time_since_epoch())
+      .count();
+}
+}  // namespace
+
+/// \brief An LRU cache of string keys and arbitrary values, with configurable
+/// max item age (in seconds) and max entries.
+///
+/// This class is thread safe.
+template <typename T>
+class ExpiringLRUCache {
+ public:
+  /// A `max_age` of 0 means that nothing is cached. A `max_entries` of 0 means
+  /// that there is no limit on the number of entries in the cache (however, if
+  /// `max_age` is also 0, the cache will not be populated).
+  ExpiringLRUCache(uint64_t max_age, size_t max_entries,
+                   TimerSeconds timer_seconds = DefaultTimer)
```

Comment:
We should be able to use https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/env.h;l=161-162;drc=2944db012c1f88bc3a453003022b985269235ca2 instead of creating a new timer. `<chrono>` is discouraged, if we can avoid it.